lib/ssl/src/tls_socket.erl

```diff
@@ -112,25 +111,27 @@ accept(ListenSocket, #config{transport_info = {Transport,_,_,_,_} = CbInfo,
 	    {error, Reason}
     end.
 
-upgrade(Socket, #config{transport_info = {Transport,_,_,_,_}= CbInfo,
-			ssl = SslOptions,
-			emulated = EmOpts}, Timeout) ->
-    ok = setopts(Transport, Socket, tls_socket:internal_inet_values()),
+upgrade(client, Socket, #config{transport_info = CbInfo,
+                                ssl = SslOptions,
+                                emulated = EmOpts}, Timeout) ->
+    Transport = element(1, CbInfo),
+    ok = setopts(Transport, Socket, internal_inet_values()),
     case peername(Transport, Socket) of
 	{ok, {Host, Port}} ->
-	    try tls_gen_connection:start_fsm(client, Host, Port, Socket,
-                                             {SslOptions,
-                                              emulated_socket_options(EmOpts, #socket_options{}), undefined},
-                                             self(), CbInfo, Timeout) of
-                Result ->
-                    Result
-            catch
-                exit:{noproc, _} ->
-                    {error, ssl_not_started}
-            end;
+            start_tls_client_connection(Host, Port, Socket, SslOptions, EmOpts, CbInfo, Timeout);
 	{error, Error} ->
 	    {error, Error}
-    end.
+    end;
+upgrade(server, Socket, #config{transport_info = CbInfo,
```

Comment:
how come the previous variant of upgrade function worked for server role? did it work?

Reply:
The server upgrade worked, but not using the upgrade function, the server upgrade was implemented directly in ssl.erl calling several functions in tls_socket. It made it harder to understand what was happening and it was asymmetrical to how the client is implemented. When the logic is the same the code should have the same pattern.  And also we should keep the related logic in the relevant modules (as we have discussed earlier)  We can continue to work on when it makes sense to have client and server separate modules, but it does not make sense in all situations, I think for some modules it makes more sense with a parameter.